libraries/plugins/debug_node/plugin.cpp

```diff
@@ -0,0 +1,671 @@
+        }
+    }
+    else {
+        while( db.head_block_time() < head_block_time ) {
+            new_blocks += debug_generate_blocks( debug_key, 1 );
+        }
+    }
+    return new_blocks;
+}
+
+uint32_t plugin::plugin_impl::debug_push_blocks(
+    std::string src_filename,
+    uint32_t count,
+    bool skip_validate_invariants
+) {
+    if( count == 0 ) {
+        return 0;
+    }
+
+    fc::path src_path = fc::path( src_filename );
+    fc::path index_path = fc::path( src_filename + ".index" );
+
+    if( fc::exists(src_path) && fc::exists(index_path) &&
+        !fc::is_directory(src_path) && !fc::is_directory(index_path)
+    ) {
+        ilog( "Loading ${n} from block_log ${fn}", ("n", count)("fn", src_filename) );
+        idump( (src_filename)(count)(skip_validate_invariants) );
+        golos::chain::block_log log;
+        log.open( src_path );
+        uint32_t first_block = database().head_block_num()+1;
+        uint32_t skip_flags = golos::chain::database::skip_nothing;
+
+        if( skip_validate_invariants ) {
+            skip_flags = skip_flags | golos::chain::database::skip_validate_invariants;
+        }
+
+        for( uint32_t i=0; i<count; i++ ) {
+            //fc::optional< chain::signed_block > block = log.read_block( log.get_block_pos( first_block + i ) );
+            uint64_t block_pos = log.get_block_pos( first_block + i );
+            if( block_pos == golos::chain::block_log::npos ) {
+                wlog( "Block database ${fn} only contained ${i} of ${n} requested blocks", ("i", i)("n", count)("fn", src_filename) );
+                return i ;
+            }
+
+            decltype( log.read_block(0) ) result;
+
+            try {
+                result = log.read_block( block_pos );
+            }
+            catch( const fc::exception& e ) {
+                elog( "Could not read block ${i} of ${n}", ("i", i)("n", count) );
+                continue;
+            }
+
+            try{
+                database().push_block( result.first, skip_flags );
+            }
+            catch( const fc::exception& e ) {
+                elog( "Got exception pushing block ${bn} : ${bid} (${i} of ${n})", ("bn", result.first.block_num())("bid", result.first.id())("i", i)("n", count) );
+                elog( "Exception backtrace: ${bt}", ("bt", e.to_detail_string()) );
+            }
+        }
+        ilog( "Completed loading block_database successfully" );
+        return count;
+    }
+    return 0;
+}
+
+fc::optional< protocol::signed_block > plugin::plugin_impl::debug_pop_block() {
+    auto & db = database();
+    return db.fetch_block_by_number( db.head_block_num() );
+}
+
+witness_schedule_object plugin::plugin_impl::debug_get_witness_schedule( ) {
+    auto & db = database();
+    return db.get( golos::chain::witness_schedule_id_type() );
+}
+
+// debug_get_hardfork_property_object_r plugin::plugin_impl::debug_get_hardfork_property_object(debug_get_hardfork_property_object_a & args) {
```

Comment:
Why this method is commented?

Reply:
@afalaleev, it's commented because there is no `api_hardfork_property_object' in golos, as it is in steem. The only similar thing we have: `hardfork_property_object`.

Reply:
I think your answer is ok. But add this information to source code with tag TODO or FIXME. Because in future it will be not clear why these lines were commented.

Reply:
Ok. Fixed